File: backend/src/main/java/com/autocare/budget/service/BudgetService.java
package com.autocare.budget.service;

import com.autocare.budget.dto.BudgetRequestDTO;
import com.autocare.budget.dto.BudgetResponseDTO;
import com.autocare.budget.entity.Budget;
import com.autocare.budget.entity.BudgetStatus;
import com.autocare.budget.repository.BudgetRepository;
import com.autocare.client.service.ClientService;
import com.autocare.serviceorder.entity.ServiceOrder;
import com.autocare.serviceorder.service.ServiceOrderService;
import com.autocare.shared.exception.BusinessException;
import com.autocare.shared.exception.ResourceNotFoundException;
import com.autocare.vehicle.service.VehicleService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class BudgetService {

    private final BudgetRepository budgetRepository;
    private final ClientService clientService;
    private final VehicleService vehicleService;
    private final ServiceOrderService serviceOrderService;

    @Transactional
    public BudgetResponseDTO create(BudgetRequestDTO request) {
        var client = clientService.findById(request.getClientId());
        var vehicle = vehicleService.findById(request.getVehicleId());

        ServiceOrder serviceOrder = null;
        if (request.getServiceOrderId() != null) {
            serviceOrder = serviceOrderService.findById(request.getServiceOrderId());
        }

        String budgetNumber = generateBudgetNumber();

        Budget budget = Budget.builder()
                .budgetNumber(budgetNumber)
                .client(client)
                .vehicle(vehicle)
                .serviceOrder(serviceOrder)
                .description(request.getDescription())
                .totalAmount(request.getTotalAmount())
                .validUntil(request.getValidUntil())
                .status(BudgetStatus.PENDENTE)
                .build();

        budget = budgetRepository.save(budget);
        return toResponseDTO(budget);
    }

    @Transactional
    public BudgetResponseDTO approve(String id) {
        Budget budget = findById(id);

        if (budget.getStatus() != BudgetStatus.PENDENTE) {
            throw new BusinessException("Apenas orçamentos pendentes podem ser aprovados");
        }

        budget.setStatus(BudgetStatus.APROVADO);
        budget = budgetRepository.save(budget);
        return toResponseDTO(budget);
    }

    @Transactional
    public BudgetResponseDTO reject(String id) {
        Budget budget = findById(id);

        if (budget.getStatus() != BudgetStatus.PENDENTE) {
            throw new BusinessException("Apenas orçamentos pendentes podem ser recusados");
        }

        budget.setStatus(BudgetStatus.RECUSADO);
        budget = budgetRepository.save(budget);
        return toResponseDTO(budget);
    }

    public BudgetResponseDTO findByIdResponse(String id) {
        return toResponseDTO(findById(id));
    }

    public List<BudgetResponseDTO> findAll() {
        return budgetRepository.findAll()
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public List<BudgetResponseDTO> findByClient(String clientId) {
        return budgetRepository.findByClientId(clientId)
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public List<BudgetResponseDTO> findByStatus(BudgetStatus status) {
        return budgetRepository.findByStatus(status)
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public Budget findById(String id) {
        return budgetRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Orçamento não encontrado com ID: " + id));
    }

    private String generateBudgetNumber() {
        String prefix = "BUD";
        String year = String.valueOf(LocalDateTime.now().getYear());
        String sequence = String.format("%06d", budgetRepository.count() + 1);
        return prefix + year + sequence;
    }

    private BudgetResponseDTO toResponseDTO(Budget budget) {
        return BudgetResponseDTO.builder()
                .id(budget.getId())
                .budgetNumber(budget.getBudgetNumber())
                .clientId(budget.getClient().getId())
                .clientName(budget.getClient().getName())
                .vehicleId(budget.getVehicle().getId())
                .vehicleInfo(budget.getVehicle().getBrand() + " " +
                        budget.getVehicle().getModel() + " (" +
                        budget.getVehicle().getPlate() + ")")
                .serviceOrderId(budget.getServiceOrder() != null ? budget.getServiceOrder().getId() : null)
                .description(budget.getDescription())
                .totalAmount(budget.getTotalAmount())
                .status(budget.getStatus())
                .validUntil(budget.getValidUntil())
                .createdAt(budget.getCreatedAt())
                .build();
    }
}

File: backend/src/budget/budget.service.ts
import { Injectable } from "@nestjs/common";
import { BudgetStatus } from "@prisma/client";
import { PrismaService } from "../prisma/prisma.service";
import { BusinessException, ResourceNotFoundException } from "../common/exceptions";
import { ClientService } from "../client/client.service";
import { VehicleService } from "../vehicle/vehicle.service";
import { ServiceOrderService } from "../service-order/service-order.service";
import { BudgetRequestDto } from "./dto/budget.dto";

@Injectable()
export class BudgetService {
  constructor(
    private prisma: PrismaService,
    private clientService: ClientService,
    private vehicleService: VehicleService,
    private serviceOrderService: ServiceOrderService,
  ) {}

  async create(dto: BudgetRequestDto) {
    const client = await this.clientService.findById(dto.clientId);
    const vehicle = await this.vehicleService.findById(dto.vehicleId);
    if (dto.serviceOrderId) {
      await this.serviceOrderService.findById(dto.serviceOrderId);
    }

    const budgetNumber = await this.generateBudgetNumber();

    const budget = await this.prisma.budget.create({
      data: {
        budgetNumber,
        clientId: client.id,
        vehicleId: vehicle.id,
        serviceOrderId: dto.serviceOrderId,
        description: dto.description,
        totalAmount: dto.totalAmount,
        validUntil: dto.validUntil ? new Date(dto.validUntil) : undefined,
        status: BudgetStatus.PENDENTE,
      },
      include: { client: true, vehicle: true },
    });
    return this.toResponse(budget);
  }

  async approve(id: string) {
    const budget = await this.findById(id);
    if (budget.status !== BudgetStatus.PENDENTE) {
      throw new BusinessException("Apenas orçamentos pendentes podem ser aprovados");
    }
    const updated = await this.prisma.budget.update({
      where: { id },
      data: { status: BudgetStatus.APROVADO },
      include: { client: true, vehicle: true },
    });
    return this.toResponse(updated);
  }

  async reject(id: string) {
    const budget = await this.findById(id);
    if (budget.status !== BudgetStatus.PENDENTE) {
      throw new BusinessException("Apenas orçamentos pendentes podem ser recusados");
    }
    const updated = await this.prisma.budget.update({
      where: { id },
      data: { status: BudgetStatus.RECUSADO },
      include: { client: true, vehicle: true },
    });
    return this.toResponse(updated);
  }

  async findByIdResponse(id: string) {
    return this.toResponse(await this.findByIdWithRelations(id));
  }

  async findAll() {
    const budgets = await this.prisma.budget.findMany({
      include: { client: true, vehicle: true },
      orderBy: { createdAt: "desc" },
    });
    return budgets.map((b) => this.toResponse(b));
  }

  async findByClient(clientId: string) {
    const budgets = await this.prisma.budget.findMany({
      where: { clientId },
      include: { client: true, vehicle: true },
    });
    return budgets.map((b) => this.toResponse(b));
  }

  async findByStatus(status: BudgetStatus) {
    const budgets = await this.prisma.budget.findMany({
      where: { status },
      include: { client: true, vehicle: true },
    });
    return budgets.map((b) => this.toResponse(b));
  }

  private async findById(id: string) {
    const budget = await this.prisma.budget.findUnique({ where: { id } });
    if (!budget) {
      throw new ResourceNotFoundException(`Orçamento não encontrado com ID: ${id}`);
    }
    return budget;
  }

  private async findByIdWithRelations(id: string) {
    const budget = await this.prisma.budget.findUnique({
      where: { id },
      include: { client: true, vehicle: true },
    });
    if (!budget) {
      throw new ResourceNotFoundException(`Orçamento não encontrado com ID: ${id}`);
    }
    return budget;
  }

  // Mesma estratégia de contador atômico usada em ServiceOrderService.
  private async generateBudgetNumber(): Promise<string> {
    const counter = await this.prisma.counter.upsert({
      where: { name: "budget" },
      update: { value: { increment: 1 } },
      create: { name: "budget", value: 1 },
    });
    const year = new Date().getFullYear();
    const sequence = String(counter.value).padStart(6, "0");
    return `BUD${year}${sequence}`;
  }

  private toResponse(budget: any) {
    return {
      id: budget.id,
      budgetNumber: budget.budgetNumber,
      clientId: budget.clientId,
      clientName: budget.client?.name,
      vehicleId: budget.vehicleId,
      vehicleInfo: budget.vehicle ? `${budget.vehicle.brand} ${budget.vehicle.model} (${budget.vehicle.plate})` : undefined,
      serviceOrderId: budget.serviceOrderId,
      description: budget.description,
      totalAmount: budget.totalAmount,
      status: budget.status,
      validUntil: budget.validUntil,
      createdAt: budget.createdAt,
    };
  }
}
